### ISSUE TITLE
Support for OSCAL as configuration input

### ISSUE THREAD

**Issue by drsm79** [ComplianceAsCode/auditree-framework#17]
## Overview

NISTs [OSCAL](https://pages.nist.gov/OSCAL/documentation/) is a public standard used for describing controls, implmentations, systems and their assessment. We have done IBM internal POC work to confirgure Auditree from OSCAL files. This is about how to bring that into the open source project.

## Requirements

1. Read standards compliant OSCAL file (model tbc - see below) to configure Auditree.
1. OSCAL input should be stored as evidence for consistency
1. identifying controls implemented via a fetcher/check should be simple & in-keeping with the spirit of OSCAL

## Approach

We have a few potential approaches, depending on how & where in the life cycle we want to introduce OSCAL. There may be more than one place it is used. This issue is to discuss & form a plan to go forward.

### Component definition

Our current POC uses a [component definition](https://pages.nist.gov/OSCAL/documentation/schema/implementation-layer/component/) to turn on/off fetchers/checks. It does this by scanning the file for `control-implementations` where `type` is `monitoring`, and then scanning the properties of the implementation for a `auditree_{fetcher,check}_path`. This is then "grafted" onto the configuration repository to produce a full config. 

To facilitate this, running auditree becomes a three step process:

1. We retrieve the component defintion via a fetcher, placing it in the locker
2. We run our configuration script to make the appropriate `controls.json` from the component defintion
3. We run the complete configuration as normal

This has some useful properties:

- combining the configuration in the component defintion with the one from the config repo happens "naturally"
- input OSCAL is stored as evidence
- the fetchers are "constrained" only by their configuration/permissions - e.g. we can "discover" inventory

### Considerations

- configuration repositories are likely more easily managed than OSCAL, so should at least be used as a baseline

## Security and Privacy

_Provide the impact on security and privacy as it relates to the completion of
this issue.  This level of detail may not be available at the time of
issue creation and can be completed at a later time.  N/A if not applicable._

## Test Plan

_Provide the test process that will be followed to adequately verify that the
approach above satisfies the requirements provided.  This level of detail may
not be available at the time of issue creation and can be completed at a later
time._

**Comment by butler54** [comment]
### Some thoughts on the comments above:
1) In terms of the steps above under the assessment plan I think step 2.1 would be validate the assessment plan. 

Assessment plan validation in this context would be non-trivial - in that you may want to validate the underlying integrity of the associated SSP / component definition** / profile and potentially underlying catalogs.

I think this will be important for a number of reasons.

1. While ideally all OSCAL artifacts will be under version control there is a decent chance some may not be (source dependent). This would give a locker so overall state of a given system could be verified and tracked (think potential changes introduced into the profile and/or a catalog in response to an audit). 
2. Even under version control it is unlikely other OSCAL artifacts will have the same degree of integrity checking as what auditree can provide.

If this is the case the validation may make sense to sit in an external tool (perhaps running in conjunction with plant).

If the above statement is a given - I think it would make sense to run the validation as an external tool.

### Configuring auditree or not from OSCAL.
- The principle challenge in use of OSCAL is the (despite OSCAL's complexity) relatively simple structure which is supported by the parameters and/or property values: Only string values

- At a first pass this seems extreme, however, OSCAL provides, with each value, 3 scoping / constraining elemenents: `ns`, `class` and `name`
- With these constrains I believe we could configure most current scenarios for auditree. `name` none of the above elements are required to be unique so lists could occur through group by style operation.


In terms of config vs OSCAL:
- I would adopt an approach where:
- OSCAL is enabled or disabled by a special flag (.oscal file with a property pointing to the assessment plan??)
- If this is the case the configuration *will* be overwritten by config generated from OSCAL
- This step may or may not be a independent pre-processing tool running in the CICD pipeline.
- Otherwise auditree proceeds as expected.





** in the case where a full SSP exists a component definition is unlikely to be needed to generated / used by Auditree

**Comment by drsm79** [comment]
re. validation - yeah, I think a check (or set of checks) would make sense. Could do a few different things:

- is the OSCAL valid (passes schema verification)?
- is there at least one fetcher/check in the OSCAL?
- are all the fetchers/checks in the OSCAL available?
- are all the fetchers/checks in the OSCAL running (e.g. configured correctly, not passing)?

I think before we worry about flags etc (I think there are clear ways to add the functionality) the question of whether we should be configuring a run or more of the whole system from OSCAL (e.g. tekton run frequency). That might point to 'built into framework' or 'separate tool' a bit more clearly.

FWIW I think the OSCAL could replace the `controls.json`, and be provided as an alternate switch to that, possibly allowing multiple (depending on which model is used - e.g. run my SOC2 assessment plan, and my ISO assessment plan, and my PCI assessment plan...). But, that's presupposing the interaction a bit.

**Comment by butler54** [comment]
I'll add more detail when i'm thinking clearly: The biggest issue I see here is operational / governance segregation. I'm wondering whether OSCAL should configure the objectives of auditree (aka the checks) and the associated fetchers; credentials should be delegated as an implementation detail for auditree.

Appreciate your thoughts - will reply my day time.

**Comment by drsm79** [comment]
Definitely think creds should be held elsewhere:

>Credential management will be a thing, especially in the more complete integration. Obviously creds should not be placed in the OSCAL files, but maybe names/references to them could be (to then be retrieved from Vault or sealed secrets).

Agree there's a scope/ease question around other parameters. Should OSCAL be used just to configure which fetchers/checks are run, with the config repo holding the actual configuration (which is a bit weird) or should it configure the fetchers/checks fully? If the latter, does that make updates awkward?

**Comment by butler54** [comment]
So the concern I would have is that i'm not sure how often an organisation will want to mess with the upstream OSCAL object. 

For example if we assume for a second the `profile` is storing the configuration of parameters - I would expect that this could become a velocity issues as I would imagine profile changes may need quite rigorous review. (i.e. I think everything in OSCAL could get very awkward).

An in-between might be that from OSCAL you *may* refer to specific URI's for configuration information for a fetch - which might allow you to keep the content bundled with a profile - or let it default to standard auditree locations. This is getting even messier thought.

**Comment by drsm79** [comment]
Right, so some kind of reference ("my configuration is in git://foo/bar/config.git") and version?

**Comment by butler54** [comment]
I think so. To me that makes sense. In the end at some point you move from 'Governance' land to 'ops' land. I think this is that method. It also allows us to hide 'non-compliance relevant' configuration to keep the OSCAL cleaner.
